### PR TITLE
Use types and typeid properties consistent with HOOMD

### LINKED ISSUE #137: Use types and typeid properties consistent with HOOMD
## Description

The properties for type names and ids are inconsistent with HOOMD, and within the Trajectory / Frame APIs.

|  API     | HOOMD/GSD                                     | garnett Frame                                       | garnett Trajectory                                                     |
|----------|-----------------------------------------------|-----------------------------------------------------|------------------------------------------------------------------------|
| type     | N/A                                           | N/A                                                 | array of N_types strings, ['A', 'B', 'C']                              |
| types    | array of N_types  strings, ['A', 'B', 'C']    | array of N_particles  strings, ['A', 'A', 'B', ...] | array of (N_frames, N_particles)  strings, [['A', 'A', 'B', ...], ...] |
| typeid   | array of N_particles indices,  [0, 0, 1, ...] | N/A                                                 | N/A                                                                    |
| type_ids | N/A                                           | N/A                                                 |                                                                        |

## Proposed solution

Refactor in reader classes:
- The type id (unsigned integer) should be the "fundamental" identifier, not the type name string

Refactor in Trajectory:
- `types` should be removed
- `type` should be removed
- `types` should be a list of lists of strings with unique types for each frame (N_frames, N_types)
- `type_ids` should become `typeid`

Refactor in Frame:
- `types` should be a list of type name strings
- `typeid` should be an array of integer type indices

## Additional context

N/A

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR rewrites the internals for `types` and `typeid` to match HOOMD's conventions.

## Motivation and Context
Resolves: #137.

## How Has This Been Tested?
The meanings of the type-related properties were changed, so I changed the expected behavior of the tests to match.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/garnett/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [credits](https://github.com/glotzerlab/garnett/blob/master/doc/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/garnett/blob/master/changelog.rst).

### CODE REVIEW

File: garnett/dcdfilereader.py
Comment (bdice) [inline garnett/dcdfilereader.py]: It's not clear to me why the `copy.copy` was required before. It may not be needed?

File: garnett/dcdfilereader.py
Comment (bdice) [inline garnett/dcdfilereader.py]: Maybe we need to `copy.copy` the fields that come from the template frame?

**Comment by bdice** [comment]
@glotzerlab/garnett-maintainers This should be ready for review.

The `types` and `typeid` properties are now a consistent source of truth for type information.

`types` looks like `['A', 'B', 'C']`. Strings with each type name, of length T (number of types, type `str`).
`typeid` looks like `[0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2]`. It's an array of indices of length N (number of particles, type `np.uint`) corresponding to each particle's type.

Because shape definitions are closely linked to our handling of types, I also re-wrote support for `type_shapes` and `shapedef` in this PR. The primary source of truth for shapes is now `type_shapes`. The `shapedef` attribute (which is an `OrderedDict`) was retained for compatibility but it's not fully functional. Setting shape definitions with `frame.shapedef['A'] = sphere_object` won't work, but setting shape definitions with `frame.shapedef = {'A': sphere_object}` will work. I am not sure how to improve on this without significantly complicating that code. I think we should deprecate/remove `shapedef` and encourage the use of `type_shapes` in its place.

Tests were only edited for two reasons:
- The meanings of the type-related properties were changed, so I changed the expected behavior of the tests to match.
- `flake8` is now applied to tests as well on CI.

Note: We were previously running flake8 only on the `garnett/` directory on CI, but the tests directory was not excluded in the `.flake8` config file. That meant that `flake8` would fail locally and pass on CI, which isn't ideal. I did not want to separate these changes since there were not many changes needed and I was offline while doing this development work (it's slightly harder to manage PRs/branches while offline). If you feel strongly about that, I can revert that and separate it into a new PR.

File: doc/quickstart.rst
Comment (lyrivera) [inline doc/quickstart.rst]: Fix `type_ids` in the list above.

File: changelog.rst
Comment (lyrivera) [inline changelog.rst]: Add deprecation note for the `shapedef` property. Maybe also mention how it works from now on until its removal.

File: README.md
Comment (lyrivera) [inline README.md]: Add `traj.N` and `traj.typeid` to the list.

**Review by lyrivera — COMMENTED** [review]
@bdice Thanks for this work. Left a few comments below. @klywang This is a big refactor so it would be nice if you could also provide a review.

File: garnett/gsdhoomdfilewriter.py
Comment (bdice) [reply garnett/gsdhoomdfilewriter.py]: Yes, that was dropped in favor of writing `type_shapes` in #127. I mentioned it in a note on that PR description. Otherwise we could be writing HPMC state into a DEM simulation, for example. I added a specific note in the changelog.

File: garnett/posfilewriter.py
Comment (bdice) [reply garnett/posfilewriter.py]: This rewritten code may be less robust in a few edge cases where there are undefined types, or more type names than type_shapes, etc. It handles those cases less well because the specification with `type_shapes` is a little more rigorously defined. I think we can accept it as-is if tests pass, and handle those edge cases if users need them. I would suspect those would only be seen in very old pre-HOOMD POS files.

File: garnett/trajectory.py
Comment (bdice) [reply garnett/trajectory.py]: My solution is less than ideal. We can handle `float32` or `float64` in garnett, and I'm just trying to detect that. I don't think adding a check for `dtype is not None` is helpful because all of the data types are defined at the top of this file. A `None` datatype would concern me here, but I don't think it's possible to have that.

Comment (lyrivera) [reply garnett/trajectory.py]: I only mentioned it because the `dtype=None` is de default argument for the function. However, now I see that  the function is called only once and a proper argument is always passed. Perhaps removing the default `None` would avoid confusion in the future.

File: garnett/gsdhoomdfilewriter.py
Comment (lyrivera) [reply garnett/gsdhoomdfilewriter.py]: Thanks for the reminder. It totally slipped my mind.

File: garnett/posfilewriter.py
Comment (lyrivera) [reply garnett/posfilewriter.py]: What I meant to ask was why were you using the `shapedef` property instead of the `type_shapes`, since `shapedef` will be removed in the future. But I can see your point now.

**Comment by bdice** [comment]
@lyrivera I'm not sure what's left to do on this -- if there is further work to be done, could I do it in a follow-up PR? There are a few PRs waiting on this so we shouldn't delay it longer than necessary.

**Comment by lyrivera** [comment]
> @lyrivera I'm not sure what's left to do on this -- if there is further work to be done, could I do it in a follow-up PR? There are a few PRs waiting on this so we shouldn't delay it longer than necessary.

@bdice I saw you marked most of the comments I left as resolved but I don't see any further changes. Did you push your latest work? If not, you can certainly address them in a follow-up PR.

**Comment by bdice** [comment]
@lyrivera Oh. Right. I didn't push. 😅 I can't find which computer I made these changes on at the moment. I'll follow up once the changes are pushed.

**Comment by lyrivera** [comment]
> @lyrivera Oh. Right. I didn't push. 😅 I can't find which computer I made these changes on at the moment. I'll follow up once the changes are pushed.

I knew something was wrong lol. No worries just request my review again once you push.

File: garnett/trajectory.py
Comment (bdice) [reply garnett/trajectory.py]: I applied this change to `typeid`. I'm not sure if we can use the same validation logic for `type_shapes` and `types`, since they're non-numeric.

Comment (lyrivera) [reply garnett/trajectory.py]: Makes sense. Perhaps in the future we can slightly modify the validation to check non-numeric arrays but it's not critical now.

**Review by lyrivera — COMMENTED** [review]
Left one more comment to consider. Once its addressed I will approve. Thanks!

**Comment by bdice** [comment]
@lyrivera I think this is probably ready.

File: garnett/dcdfilereader.py
Comment (bdice) [reply garnett/dcdfilereader.py]: I fixed the dtype. @klarh says it looks like that info is encoded once at the top of the file, not per-frame. I'm fine with making the constant-N assumption and fixing it later, if it's ever needed.

Comment (lyrivera) [reply garnett/dcdfilereader.py]: Sounds good.

File: garnett/posfilewriter.py
Comment (lyrivera) [inline garnett/posfilewriter.py]: I'm OK with erroring out if there are not enough shape definitions but it should probably be mentioned in the `changelog` since its a change from the expected behavior.

**Review by lyrivera — APPROVED** [review]
Left one last minor comment, but other than that, looks good to go! Thanks again!

**Pull request merged**
